### PR TITLE
[GLUTEN-5965][VL] Support pushdown "not in" to scan node

### PR DESCRIPTION
**Pull request by WangGuangxin**
## What changes were proposed in this pull request?

Support pushdown `not in` to scan node

(Fixes: \#5965)

## How was this patch tested?

UT

### CODE REVIEW

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 1]
```diff
@@ -1449,10 +1450,14 @@ connector::hive::SubfieldFilters SubstraitToVeloxPlanConverter::createSubfieldFi
       auto expr = scalarFunction.arguments()[0].value();
       if (expr.has_scalar_function()) {
         // Set its child to filter info with reverse enabled.
-        setFilterInfo(scalarFunction.arguments()[0].value().scalar_function(), inputTypeList, columnToFilterInfo, true);
+        setFilterInfo(expr.scalar_function(), inputTypeList, columnToFilterInfo, true);
+      } else if (expr.has_singular_or_list()) {
+        // Since currently only integer and string types in "IN" expression
+        // can be pushed down, Options types should be checked
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:1456]: > Options types should be checked

Could you remind me where the check for types is added?

File: cpp/velox/substrait/SubstraitToVeloxPlan.h [hunk 2]
```diff
@@ -364,9 +364,11 @@ class SubstraitToVeloxPlanConverter {
...
     }
 
-    // Set a value for the not(equal) condition.
-    void setNotValue(const std::optional<variant>& notValue) {
-      notValue_ = notValue;
+    // Set a list of values to be used in the push down of 'not in' expression.
+    void setNotValue(const std::vector<variant>& notValue) {
+      for (const auto& value : notValue) {
+        notValues_.emplace_back(value);
+      }
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.h:371]: nit: suggestion on the namings: notValue -> notValues, setNotValue -> setNotValues

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 3]
```diff
@@ -1577,26 +1582,28 @@ bool SubstraitToVeloxPlanConverter::canPushdownNot(
...
+    bool isFieldOrWithLiteral = fieldOrWithLiteral(notArg.value().scalar_function().arguments(), fieldIdx);
 
-  if (supportedNotFunctions.find(functionName) != supportedNotFunctions.end() && isFieldOrWithLiteral &&
-      rangeRecorders.at(fieldIdx).setCertainRangeForFunction(functionName, true /*reverse*/)) {
-    return true;
+    if (supportedNotFunctions.find(functionName) != supportedNotFunctions.end() && isFieldOrWithLiteral &&
+        rangeRecorders.at(fieldIdx).setCertainRangeForFunction(functionName, true /*reverse*/)) {
+      return true;
+    }
+    return false;
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:1603]: This return seems to be not needed as the else branch will return false anyway.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 4]
```diff
@@ -2118,44 +2107,25 @@ void SubstraitToVeloxPlanConverter::constructSubfieldFilters(
...
-    // Handle not(equal) filter.
-    if (filterInfo.notValue_) {
-      variant notVariant = filterInfo.notValue_.value();
-      createNotEqualFilter<KIND, FilterType>(notVariant, filterInfo.nullAllowed_, colFilters);
+    // Handle not in filter.
+    if (filterInfo.notValues_.size() > 0) {
+      setInFilter<KIND>(filterInfo.notValues_, filterInfo.nullAllowed_, true, inputName, filters);
       // Currently, Not-equal cannot coexist with other filter conditions
       // due to multirange is in 'OR' relation but 'AND' is needed.
       VELOX_CHECK(rangeSize == 0, "LowerBounds or upperBounds conditons cannot be supported after not-equal filter.");
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:2128]: Could you also update the comment and error message as this branch handles more cases than not-equal?

File: backends-velox/src/test/scala/org/apache/gluten/execution/TestOperator.scala [hunk 5]
```diff
@@ -156,6 +156,64 @@ class TestOperator extends VeloxWholeStageTransformerSuite {
...
+      "select l_orderkey from lineitem " +
+        "where l_partkey not in (1552, 674) and l_partkey >= 1552") {
+      checkGlutenOperatorMatch[FileSourceScanExecTransformer]
+    }
+
+    // mix not-in with in
+    runQueryAndCompare(
+      "select l_orderkey from lineitem " +
+        "where l_partkey not in (1552, 674) and l_partkey in (1552)") {
+      checkGlutenOperatorMatch[FileSourceScanExecTransformer]
```
Comment (rui-mo) [inline backends-velox/src/test/scala/org/apache/gluten/execution/TestOperator.scala:213]: Could we add a test for or relation like `l_partkey not in (1552, 674) or l_partkey in (1552)`?

**Review by rui-mo — COMMENTED** [review]
Thanks for your work. Just added several comments.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc (hunk 1 above)
Comment (WangGuangxin) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc]: Sorry, the comment is useless

File: cpp/velox/substrait/SubstraitToVeloxPlan.h (hunk 2 above)
Comment (WangGuangxin) [reply cpp/velox/substrait/SubstraitToVeloxPlan.h:371]: done

**Comment by WangGuangxin** [comment]
@rui-mo comments resolved and all ut passed, please help review again, thanks

File: cpp/velox/substrait/SubstraitToVeloxPlan.h [hunk 6]
```diff
@@ -479,13 +494,14 @@ class SubstraitToVeloxPlanConverter {
   template <TypeKind KIND, typename FilterType>
   void createNotEqualFilter(variant notVariant, bool nullAllowed, std::vector<std::unique_ptr<FilterType>>& colFilters);
 
-  /// Create a values range to handle in filter.
-  /// variants: the list of values extracted from the in expression.
+  /// Create a values range to handle (not) in filter.
+  /// variants: the list of values extracted from the (not) in expression.
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.h:498]: Is the behavior to create `in` or `not in` controlled by the variable `negated`?

**Review by rui-mo — COMMENTED** [review]
Thanks. Looks good overall.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h (hunk 6 above)
Comment (WangGuangxin) [reply cpp/velox/substrait/SubstraitToVeloxPlan.h:498]: yes, add comments for this param

**Review by rui-mo — APPROVED** [review]
Thanks!

**Pull request merged**